### ISSUE TITLE
[API Proposal]: Use TimeSpan everywhere we use an int for seconds, milliseconds, and timeouts (Group 2/3)

### ISSUE THREAD

**Issue by deeprobin** [dotnet/runtime#67156]
### Background and motivation

[As discussed with @danmoseley](https://github.com/dotnet/runtime/issues/14336#issuecomment-1065295310) this issue serves as a follow up for #14336 and is meant to represent the 2nd group

### API Proposal

```cs
namespace System.IO {
    public abstract class Stream {
       public TimeSpan ReadTimeoutTimeSpan { get; set; }
       public TimeSpan WriteTimeoutTimeSpan { get; set; }
    }
}

namespace System.IO.Ports {
    public class SerialPort {
       public TimeSpan ReadTimeoutTimeSpan { get; set; }
       public TimeSpan WriteTimeoutTimeSpan { get; set; }
    }
}

namespace System.Media {
    public class SoundPlayer {
       public TimeSpan LoadTimeoutTimeSpan { get; set; } 
    }
}

namespace System.Net.NetworkInformation {
   public static class NetworkInformationTimeSpanExtensions {
       public static TimeSpan GetPacketReassemblyTimeout(this IPGlobalStatistics statistics);
       public static TimeSpan GetMaximumTransmissionTimeout(this TcpStatistics statistics);
       public static TimeSpan GetMinimumTransmissionTimeout(this TcpStatistics statistics);
   }
}

namespace System.Net.Sockets {
    public class Socket {
       public TimeSpan ReceiveTimeoutTimeSpan { get; set; }
       public TimeSpan ReceiveTimeoutTimeSpan { get; set; }
    }

    public class TcpClient {
       public TimeSpan ReceiveTimeoutTimeSpan { get; set; }
       public TimeSpan SendTimeoutTimeSpan { get; set; }
    }
}

namespace System.Timers {
    public class Timer {
       public TimeSpan IntervalTimeSpan { get; set; }
    }
}
```

**Comment by deeprobin** [comment]
@danmoseley Should I outsource the SqlClient specific stuff to the [SqlClient repo](https://github.com/dotnet/SqlClient/) or will the API still be discussed in this repo?

**Comment by GSPP** [comment]
With respect to `SqlConnectionStringBuilder`: Does the SQL Server connection string even support non-second timeouts? I believe not, and this would then not be a faithful representation of a connection string. Using `TimeSpan` here would suggest a feature that isn't there, with very rough emulation (rounding to seconds).

**Comment by deeprobin** [comment]
@Wraith2 You do quite a lot in the SqlClient. Do you know anything about it?

**Comment by Wraith2** [comment]
The TDS spec is remarkably cagey about it. Timers are an implementation detail or the client or server and apart from some defaults being in seconds it makes no comment about resolution minimum or maximums.
To the best of my knowledge all current durations in the client library are expressed in seconds but there's no absolute reason that this needs to be the case. I expect that this will continue o be the case. 

Adding a timespan wouldn't be harmful and would possibly be an extension of functionality iff client libraries chose to use a higher resolution. I would not expect Microsoft.Data.SqlClient to do that for quite a long time but faster moving libraries like Npgqsl and MySqlConnector to do so if their protocols support higher resolutions. You'd probably want to ask @roji and @bgrainger

**Comment by roji** [comment]
> Using TimeSpan here would suggest a feature that isn't there, with very rough emulation (rounding to seconds).

I don't think there's some value in allow TimeSpan representations even where non-second timeouts aren't supported: there's no doubt as to the unit (where the meaning of current int properties is completely implicit), and it allows specifying timeouts in larger units, e.g. minutes (though that's not likely to be very useful).

Simply adding a DIM IDbCommand.CommandTimeoutTimeSpan doesn't seem very useful - users would need to explicitly cast to IDbCommand in order to make use of it. However, we could also add CommandTimeoutTimeSpan to DbCommand, which is the abstract base class for most ADO.NET command implementations. The default implementation could simply convert the TimeSpan to seconds (and back), and throw if sub-second intervals are specified. Providers which support sub-second intervals would be able to override the property.

What happens with connection string builders (e.g. SqlConnectionStringBuilder) is obviously provider-specific, and each provider can do what's appropriate. If such a property is added, I'd also advise throwing (and not rounding) for unsupported values (e.g. timeouts which don't round to whole second intervals), just like with DbCommand.CommandTimeoutTimeSpan.

Since System.Data.SqlClient is in maintenance-mode, I think any SqlClient-specific additions would probably make sense in Microsoft.Data.SqlClient rather than in System.Data.SqlClient.

/cc @ajcvickers

**Comment by danmoseley** [comment]
> I don't think there's some value

@roji can you clarify, I can't tell whether you mean there's value or not based on the rest of what you wrote.

**Comment by roji** [comment]
@danmoseley sorry, there was some editing happening there - I *do* think there's some value there.

**Comment by GSPP** [comment]
I'm very much for supporting sub-second timeouts (and especially getting rid of the tradition of using integers to represent time - we are not C in 1990). If such configurations are exposed, though, this should go hand in hand with actually supporting it.

I believe that SQL Server timeouts are implemented mostly as client-side timeouts. The client sends an "attention" which is a cancellation request. I'm not sure the server ever times out anything (maybe with Resource Governor or with a concurrent `KILL` command).

**Comment by Clockwork-Muse** [comment]
> Throwing in case of non-second-integral timeouts seems problematic: Time can involve floating point math and you can end up with a timeout of 2.99999999 seconds.

This is why I loathe the current api that takes `double` for adding units to `TimeSpan` (and others) - it absolutely shouldn't be.  The number should be understood to be an exact base-10 representation, and it should be parsed as such - it's represented in `TimeSpan` that way, after all.

> Querying a single small row by primary key from memory takes far below 1ms.

The query time is almost never the problem; it's usually network related (either the network itself, or things like serialization of the results).

**Comment by Wraith2** [comment]
> If the TimeSpan capability was added to IDbCommand directly, then it would seem like an appropriate trade-off to make. But then, SqlCommand maybe should implement that interface member explicitly to hide it by default.

Any additions to Microsoft.Data.SqlClient will take considerable time regardless of difficulty. There is not likely to be a net7.0 build so the earliest this would get in is likely 8 which is 18+ months away.

**Comment by Wraith2** [comment]
Being outside runtime makes that possible. Unfortunately small team with a big backlog means that anything new will take time. See the area owner's notes in https://github.com/dotnet/runtime/issues/51836#issuecomment-1012664600 for the basis of my information.

**Comment by deeprobin** [comment]
@ericstj 
Should we outsource the SqlClient specific to an external issue and mark the part for the BCL + runtime libraries as `ready-for-review`?

**Comment by ericstj** [comment]
Yes, that sounds like a good plan.  The SqlClient API requests can go to https://github.com/dotnet/SqlClient/ as you suggested before.

**Comment by deeprobin** [comment]
@ericstj See https://github.com/dotnet/SqlClient/issues/1602. 
Can you label this issue with `api-ready-for-review`?

**Comment by tarekgh** [comment]
General comment about naming here, I am not a fan of specifying the type `TimeSpan` in the property names. Should we use some other term? something like `Precise`? so, we can have something like `ReadPreciseTimeout`?

**Comment by tarekgh** [comment]
tagging some more people for awareness: @dotnet/ncl @dotnet/area-system-io @dotnet/area-microsoft-win32 @dotnet/area-system-runtime @ajcvickers @davoudeshtehari @david-engel

**Comment by mattjohnsonpint** [comment]
Question about some of the proposed APIs that are properties with setters.  What happens if there's a conflict?

For example:

```csharp
var client = new TcpClient
{
    ReceiveTimeout = 5000;
    ReceiveTimeoutTimeSpan = TimeSpan.FromMinutes(10);
}
```

Which one wins?

**Comment by tarekgh** [comment]
I guess the last set is the winner.

**Comment by Wraith2** [comment]
> there can be some truncation if we try to do the conversion. The doc would need clarify the behavior.

Channelling tanner here, the rounding/truncation method should be specified if possible. It would be confusing if some providers round up and others down or throw.

**Comment by tarekgh** [comment]
> Channelling tanner here, the rounding/truncation method should be specified if possible. It would be confusing if some providers round up and others down or throw.

All proposed APIs here are non-virtual. So, the implementation of the new APIs can decide the truncation method and we can document it. The only exception is the `System.Data.IDbCommand` though.

**Comment by ericstj** [comment]
> Question about some of the proposed APIs that are properties with setters. What happens if there's a conflict?

> I guess the last set is the winner.

Yes, that's what I would assume.  All of these are just a different view over the existing storage.  @bartonjs touched on this here https://github.com/dotnet/runtime/pull/64860#discussion_r830135382

**Comment by bartonjs** [comment]
The general consensus is that the need for two properties (and the TimeSpan suffix) is probably as much of a detractor to the user experience as the benefit of having clear units for the timeout.

Since this is only for a small number of types, and there's not a whole lot of benefit, we think the answer is to just not make the same mistake in the future and let what we have be what we have.

The individual area owners are highly encouraged to make sure that the `<summary>` text for each of these properties makes the unit very clear.
